pkg/udf/forward/forward.go

```diff
@@ -409,32 +409,58 @@ func (isdf *InterStepDataForward) processConcurrentMap(ctx context.Context, data
 // response for all of them.
+// if there is an error it will do a retry and will return when
+// - if there is a success while retrying
+// - if shutdown has been initiated.
+// - if context is cancelled
 func (isdf *InterStepDataForward) processBatchMessages(ctx context.Context, dataMessages []*isb.ReadMessage) []isb.ReadWriteMessagePair {
 	concurrentUDFProcessingStart := time.Now()
-	udfResults, err := isdf.opts.batchMapUdfApplier.ApplyBatchMap(ctx, dataMessages)
-	if err != nil {
-		// In case of an error received while processing the UDF call, we would do not handle partial failures,
-		// and hence would want to replay the whole batch currently.
-		// For achieving this, no-ack all the messages in the batch to force an early re-read from the ISB
-		// And then restart the numa container. Though we cannot enforce from the ISB that the same set of messages
-		// will be sent back everytime there is a replay.
-		//TODO(map-batch) - We do not have any retry mechanism currently in place. Do we want to add any before
-		// restarting?
-		// Also check why we do not have a retry mechanism in no-ack similar to ackFromBuffer
-		if ok, _ := isdf.IsShuttingDown(); ok {
-			isdf.opts.logger.Errorw("batchMapUDF.Apply, Stop called during udf processing", zap.Error(err))
-			metrics.PlatformError.With(map[string]string{metrics.LabelVertex: isdf.vertexName, metrics.LabelPipeline: isdf.pipelineName, metrics.LabelVertexType: string(dfv1.VertexTypeMapUDF), metrics.LabelVertexReplicaIndex: strconv.Itoa(int(isdf.vertexReplica))}).Inc()
+	var udfResults []isb.ReadWriteMessagePair
+	var err error
+	success := false
+udfLoop:
+	for {
+		// invoke the UDF call
+		udfResults, err = isdf.opts.batchMapUdfApplier.ApplyBatchMap(ctx, dataMessages)
+		if err != nil {
+			// check if there is a shutdown, in this case we will not retry further
+			if ok, _ := isdf.IsShuttingDown(); ok {
+				isdf.opts.logger.Errorw("batchMapUDF.Apply, Stop called during udf processing", zap.Error(err))
+				metrics.PlatformError.With(map[string]string{metrics.LabelVertex: isdf.vertexName,
+					metrics.LabelPipeline: isdf.pipelineName, metrics.LabelVertexType: string(dfv1.VertexTypeMapUDF),
+					metrics.LabelVertexReplicaIndex: strconv.Itoa(int(isdf.vertexReplica))}).Inc()
+				break udfLoop
+			}
+			isdf.opts.logger.Errorw("batchMapUDF.Apply got error during batch udf processing", zap.Error(err))
+			select {
+			case <-ctx.Done():
+				// no point in retrying if the context is cancelled
+				break udfLoop
+			case <-time.After(1 * time.Second):
+				// sleep for a second and keep retrying after that
+				continue
+			}
 		}
-		readOffsets := make([]isb.Offset, 0)
-		for _, msg := range dataMessages {
-			readOffsets = append(readOffsets, msg.ReadOffset)
+		// if no error, let's mark this as a success and break
+		success = true
+		break
+	}
+	if !success {
+		udfErr := fmt.Errorf("batchMapUDF.Apply got error during batch udf processing")
```

Comment:
The error message is too generic, making it hard for debugging. Can we create and return the `udfResultsErr` object at each break statement and add more descriptions to the error message? That way we can also remove the success flag, making the code more readable.